### ISSUE TITLE
Issue in running post quantised TFLite Model using ArmNN

### ISSUE THREAD

**Issue by kg512** [ARM-software/armnn#425]
I have retrained MobileNet v2 with lesser classes and converted to uint8 model with Post training quantization using TF1.15 (Keras APIs)

The TFLite Model works with a python code, however using it with ArmNN, leads to the following error:

Optimize graph and load the optimized graph onto a compute device...
Warning: WARNING: Layer of type Convolution2d is not supported on requested backend GpuAcc for input data type QAsymmS8 and output data type QAsymmS8 (reason: in validate src/runtime/CL/functions/CLGEMMConvolutionLayer.cpp:423: Input data type not compatible with Weights), falling back to the next backend.
Warning: ERROR: Layer of type Convolution2d is not supported on any preferred backend [GpuAcc ]

**Comment by MikeJKelly** [comment]
Hi @kratika5 

That error occurs if the weights are QSymm8 with per channel quantization (that's QSYMM8_PER_CHANNEL) and the input is not QAsymm8. In this case your input is QAsymmS8 so it's not supported right now.  

In order to use the GpuAcc you would have to retrain the model to use QASymm8 instead of QAsymmS8 or not to use per channel quantization on your weights.

Best regards,
Mike

**Comment by kratika5** [comment]
Hi Mike,

Thanks for the reply. I am not really sure, how I can do this using TFLite Converter (Tf-Keras APIs). Thus, I have posted for help on tensorflow git. If you have any guidance in that direction. Please help.

**Comment by MikeJKelly** [comment]
Hi @kratika5 

I'm afraid I don't know how you might do this on the TFLite Converter, however the ComputeLibrary team have improved the support for QAsymmS8 in the following patch: https://review.mlplatform.org/c/ml/ComputeLibrary/+/3452  

We are currently working on updating the supported version of ComputeLibrary in ArmNN and hope to have that patch supported in the next week or so. 

If you want to you could try applying that patch directly to your own computelibrary version?

Best regards,
Mike

**Comment by janeil01** [comment]
Hi @kratika5 

the patch that Mike mentioned is now supported on our armnn master branch. Your network should be fully supported if you change to that branch. Also make sure that you checkout ComputeLibrary at the new pin before you give that a go.

Good luck and let me know if you have any other issues
Jan

**Comment by kratika5** [comment]
Thank you for the replies.

I built the new release of Compute Library from https://review.mlplatform.org/admin/repos/ml%2FComputeLibrary; 
and ArmNN from https://github.com/ARM-software/armnn/tree/master.

However I am facing the following issue while doing "make" for ArmNN. I am not sure if it because I am not using the correct branch for CL and ArmNN or something else. Thank you for your help in advance.

Regards,
Kratika 

[ 17%] Building CXX object src/backends/cl/workloads/CMakeFiles/armnnClBackendWorkloads.dir/ClConcatWorkload.cpp.o
/home/odroid/armnn-tflite/armnn/src/backends/cl/workloads/ClConcatWorkload.cpp: In function ‘arm_compute::Status armnn::ClConcatWorkloadValidate(const std::vector<const armnn::TensorInfo*>&, const armnn::TensorInfo&, const armnn::OriginsDescriptor&)’:
/home/odroid/armnn-tflite/armnn/src/backends/cl/workloads/ClConcatWorkload.cpp:46:91: error: no matching function for call to ‘arm_compute::CLConcatenateLayer::validate(std::vector<arm_compute::ITensorInfo*>&, const arm_compute::TensorInfo*, size_t&)’
 m_compute::CLConcatenateLayer::validate(aclInputPtrs, &aclOutputInfo, aclAxis);
                                                                              ^
In file included from /home/odroid/armnn-tflite/armnn/src/backends/cl/workloads/ClConcatWorkload.cpp:14:0:
/home/odroid/armnn-tflite/ComputeLibrary/arm_compute/runtime/CL/functions/CLConcatenateLayer.h:97:19: note: candidate: static arm_compute::Status arm_compute::CLConcatenateLayer::validate(const std::vector<const arm_compute::ITensorInfo*>&, const arm_compute::ITensorInfo*, size_t)
     static Status validate(const std::vector<const ITensorInfo *> &inputs_vector, const ITensorInfo *output, size_t axis);
                   ^~~~~~~~
/home/odroid/armnn-tflite/ComputeLibrary/arm_compute/runtime/CL/functions/CLConcatenateLayer.h:97:19: note:   no known conversion for argument 1 from ‘std::vector<arm_compute::ITensorInfo*>’ to ‘const std::vector<const arm_compute::ITensorInfo*>&’
/home/odroid/armnn-tflite/armnn/src/backends/cl/workloads/ClConcatWorkload.cpp: In constructor ‘armnn::ClConcatWorkload::ClConcatWorkload(const armnn::ConcatQueueDescriptor&, const armnn::WorkloadInfo&)’:
/home/odroid/armnn-tflite/armnn/src/backends/cl/workloads/ClConcatWorkload.cpp:86:49: error: no matching function for call to ‘arm_compute::CLConcatenateLayer::configure(std::vector<arm_compute::ICLTensor*>&, arm_compute::ICLTensor*, size_t&)’
     layer->configure(aclInputs, &output, aclAxis);
                                                 ^
In file included from /home/odroid/armnn-tflite/armnn/src/backends/cl/workloads/ClConcatWorkload.cpp:14:0:
/home/odroid/armnn-tflite/ComputeLibrary/arm_compute/runtime/CL/functions/CLConcatenateLayer.h:74:10: note: candidate: void arm_compute::CLConcatenateLayer::configure(std::vector<const arm_compute::ICLTensor*>&, arm_compute::ICLTensor*, size_t)
     void configure(std::vector<const ICLTensor *> &inputs_vector, ICLTensor *output, size_t axis);
          ^~~~~~~~~
/home/odroid/armnn-tflite/ComputeLibrary/arm_compute/runtime/CL/functions/CLConcatenateLayer.h:74:10: note:   no known conversion for argument 1 from ‘std::vector<arm_compute::ICLTensor*>’ to ‘std::vector<const arm_compute::ICLTensor*>&’
/home/odroid/armnn-tflite/ComputeLibrary/arm_compute/runtime/CL/functions/CLConcatenateLayer.h:85:10: note: candidate: void arm_compute::CLConcatenateLayer::configure(const arm_compute::CLCompileContext&, std::vector<const arm_compute::ICLTensor*>&, arm_compute::ICLTensor*, size_t)
     void configure(const CLCompileContext &compile_context, std::vector<const ICLTensor *> &inputs_vector, ICLTensor *output, size_t axis);
          ^~~~~~~~~
/home/odroid/armnn-tflite/ComputeLibrary/arm_compute/runtime/CL/functions/CLConcatenateLayer.h:85:10: note:   candidate expects 4 arguments, 3 provided
src/backends/cl/workloads/CMakeFiles/armnnClBackendWorkloads.dir/build.make:230: recipe for target 'src/backends/cl/workloads/CMakeFiles/armnnClBackendWorkloads.dir/ClConcatWorkload.cpp.o' failed
make[2]: *** [src/backends/cl/workloads/CMakeFiles/armnnClBackendWorkloads.dir/ClConcatWorkload.cpp.o] Error 1
CMakeFiles/Makefile2:1976: recipe for target 'src/backends/cl/workloads/CMakeFiles/armnnClBackendWorkloads.dir/all' failed
make[1]: *** [src/backends/cl/workloads/CMakeFiles/armnnClBackendWorkloads.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by janeil01** [comment]
Sorry, I should have thought about this. You're on the right track. Looks like your acl version doesn't match with armnn. In your armnn directory there is a folder called scripts. In scripts there is a script (who would have guessed that) that is called get_compute_library.sh. If you open that up with an editor of your choice, I used nano, 
```
cd <rootdir>/armnn/scripts
nano get_compute_library.sh
```
you'll see on top of the file something like this:
```
# For pinning to a revision use this:
DEFAULT_CLFRAMEWORKREVISION="ad7515d231acb075a9585e52f257373b1a1b5d1f"
```
Copy the sha, in this case `ad7515d231acb075a9585e52f257373b1a1b5d1f`, then change to your acl directory and checkout the copied sha:
```
cd <rootdir>/arm_compute <-- I'm just guessing where your compute library is located, you won't be able to copy this command
git checkout ad7515d231acb075a9585e52f257373b1a1b5d1f
```
Now you can recompile arm compute library and armnn and it should work :+1:

**Comment by kratika5** [comment]
Hi,

Thank you for the reply. 

ArmNN: https://github.com/ARM-software/armnn/tree/master
ComputeLibrary: aba2f912a21487776e540724cf354a6cd8d89829 (From get_compute_library.sh)

After building these, it has alleviated the problem for CpuAcc, but the problem still persists for GpuAcc. 
The following is the error:
Import graph...
Optimize graph and load the optimized graph onto a compute device...
Info: ArmNN v21.0.0
Info: Initialization time: 5.35 ms
Warning: WARNING: Layer of type Convolution2d is not supported on requested backend GpuAcc for input data type QAsymmS8 and output data type QAsymmS8 (reason: in validate src/runtime/CL/functions/CLGEMMConvolutionLayer.cpp:423: Input data type not compatible with Weights), falling back to the next backend.
Warning: ERROR: Layer of type Convolution2d is not supported on any preferred backend [GpuAcc ]
Graph optimized...

**Comment by janeil01** [comment]
Can you show me your output for 
```
cd <root>/armnn
git log -1
```
and 
```
cd <root>/arm_compute
git log -1
```
I'd like to know what versions for armnn and acl you are working on right now.

**Comment by kratika5** [comment]
Hi,

Thank you for the reply. Here it is:
**ArmNN**: 
commit ea104f29baf384eb098215fd55fd0bdd680711c2 (HEAD -> master, tag: v20.05, origin/branches/armnn_20_05, origin/HEAD, branches/armnn_20_05, armnn_20_05)
Author: Colm Donelan <Colm.Donelan@arm.com>
Date:   Fri May 29 11:15:40 2020 +0100

    IVGCVSW-4801: Updating Doxygen project version for 20.05 release.
    
    Signed-off-by: Colm Donelan <Colm.Donelan@arm.com>
    Change-Id: I408a10610a5a91de4608b2597918445fcb5ee21c

**ComputeLibrary**
commit aba2f912a21487776e540724cf354a6cd8d89829 (HEAD)
Author: morgolock <pablo.tello@arm.com>
Date:   Tue May 5 16:28:19 2020 +0100

    COMPMID-3289: Test improvement CLGEMMMatrixMultiplyReshapedKernel.
    
    Change-Id: Ia6c2f115849889baceafaf716477456e41f96037
    Signed-off-by: morgolock <pablo.tello@arm.com>
    Reviewed-on: https://review.mlplatform.org/c/ml/ComputeLibrary/+/3186
    Comments-Addressed: Arm Jenkins <bsgcomp@arm.com>
    Tested-by: Arm Jenkins <bsgcomp@arm.com>
    Reviewed-by: Sang-Hoon Park <sang-hoon.park@arm.com>

**Comment by janeil01** [comment]
Alright, your armnn branch is not up to date. Do a git pull in armnn
```
cd <root>/armnn
git checkout master
git pull
```
Ensure that this commit is included
```
    Update ACL pin 42e57362933b7fa04f9a86297fa75ed38c5e3c97
    
    Signed-off-by: Teresa Charlin <teresa.charlinreyes@arm.com>
    Change-Id: I7c27c371c00aeb1eba61b7a0137232cca175c8dd
```
And then repeat the steps to get the correct acl version form `get_compute_library.sh`

Now it should wokr :crossed_fingers:

**Comment by kratika5** [comment]
Hi,

Unfortunately, it is still not working! Now the error is coming up for a different layer:
Warning: WARNING: Layer of type Mean is not supported on requested backend GpuAcc for input data type QAsymmS8 and output data type QAsymmS8 (reason: in validate_config src/runtime/CL/functions/CLReduceMean.cpp:86: Tensors have different quantization information), falling back to the next backend.
Warning: ERROR: Layer of type Mean is not supported on any preferred backend [GpuAcc ]
Graph optimized...

**ArmNN**:
commit 9a2d1d004473bbba25e1c9ecb0b8d6fa7fced6f1 (HEAD)
Author: Teresa Charlin <teresa.charlinreyes@arm.com>
Date:   Wed Jul 29 14:47:39 2020 +0100

    Update ACL to a179798501af2b3939f9282b2f03ef4f98471d81
    
    Signed-off-by: Teresa Charlin <teresa.charlinreyes@arm.com>
    Change-Id: Iec00568a6feaeeff816025d9cc38531090d12cba

**ComputeLibrary**:

commit a179798501af2b3939f9282b2f03ef4f98471d81 (HEAD, origin/master, origin/branches/arm_compute_20_08, origin/HEAD, master)
Author: Michele Di Giorgio <michele.digiorgio@arm.com>
Date:   Tue Jul 28 16:59:48 2020 +0100

    COMPMID-3585: Android R while_fib_n_5_quant8 failure on CpuAcc
    
    Assembly kernels do not support quantized GEMM when the multiplier is
    greater than 1 (which leads to negative result shift used for
    requantization).
    
    Change-Id: I7a766cd0d13f549d217613ca67bc952923f309de
    Signed-off-by: Michele Di Giorgio <michele.digiorgio@arm.com>
    Reviewed-on: https://review.mlplatform.org/c/ml/ComputeLibrary/+/3538
    Tested-by: Arm Jenkins <bsgcomp@arm.com>
    Reviewed-by: Georgios Pinitas <georgios.pinitas@arm.com>
    Comments-Addressed: Arm Jenkins <bsgcomp@arm.com>

**Comment by janeil01** [comment]
Hey,

that's a known issue. Acl doesn't support different quantizations for input and output in reduce mean. There is an internal ticket created already and they'll start working on it shortly. 

Sorry, I don't have a quick fix for that one. I'll let you know when acl has added support for it.

Jan

**Comment by janeil01** [comment]
Hey @kratika5 

I've got an update for this ticket.
That issue with the reduce mean operation you had should be fixed now. Try the current armnn master version if you like :)

Jan

**Comment by MikeJKelly** [comment]
Hi,
I'm closing this as I believe that the issue has been resolved. If you require any more help please open another issue. 

Best regards,
Mike
